File: src/10.graph/10.39.ts
/**
 * 不进行剪枝
 * @param candidates
 * @param target
 * @returns
 */
export function combinationSum1(
    candidates: number[],
    target: number
): number[][] {
    const ans: number[][] = [];
    const onPath: number[] = [];
    let total = 0;

    const backtrack = (r: number) => {
        if (total === target) {
            ans.push([...onPath]);
            return;
        }

        for (let i = r; i < candidates.length; i++) {
            const val = candidates[i];

            // 如果当前和大于了, 就跳过这个选择
            if (total + val > target) {
                continue;
            }

            total += val;
            onPath.push(val);

            // 可以进行重复选择, 所以不用 + 1
            backtrack(i);

            total -= val;
            onPath.pop();
        }
    };

    backtrack(0);

    return ans;
}

/**
 *
 * @param candidates
 * @param target
 * @returns
 */
export function combinationSum(
    candidates: number[],
    target: number
): number[][] {
    const ans: number[][] = [];
    const onPath: number[] = [];
    let total = 0;

    // 先排序, 大于了之后, 直接跳出循环
    candidates.sort((a, b) => a - b);

    const backtrack = (r: number) => {
        if (total === target) {
            ans.push([...onPath]);
            return;
        }

        for (let i = r; i < candidates.length; i++) {
            const val = candidates[i];

            // 如果当前和大于了, 因为排序过, 就跳过整个循环
            if (total + val > target) {
                break;
            }

            total += val;
            onPath.push(val);

            // 可以进行重复选择, 所以不用 + 1
            backtrack(i);

            total -= val;
            onPath.pop();
        }
    };

    backtrack(0);

    return ans;
}

File: src/10.graph/10.39.go
package graph

import (
	"sort"
)

func CombinationSum1(candidates []int, target int) [][]int {
	ans := [][]int{}
	onPath := []int{}
	total := 0
	var backtrack func(startIndex int)

	backtrack = func(startIndex int) {
		if total == target {
			ans = append(ans, append([]int{}, onPath...))
			return
		}

		for i := startIndex; i < len(candidates); i++ {
			val := candidates[i]

			if total+val > target {
				continue
			}

			total += val
			onPath = append(onPath, val)

			backtrack(i)

			total -= val
			onPath = onPath[:len(onPath)-1]
		}
	}

	backtrack(0)

	return ans
}

func CombinationSum(candidates []int, target int) [][]int {
	ans := [][]int{}
	onPath := []int{}
	total := 0
	var backtrack func(startIndex int)

	sort.Ints(candidates)

	backtrack = func(startIndex int) {
		if total == target {
			ans = append(ans, append([]int{}, onPath...))
			return
		}

		for i := startIndex; i < len(candidates); i++ {
			val := candidates[i]

			if total+val > target {
				break
			}

			total += val
			onPath = append(onPath, val)

			backtrack(i)

			total -= val
			onPath = onPath[:len(onPath)-1]
		}
	}

	backtrack(0)

	return ans
}
